### PR TITLE
Improvements before release

### PR DESCRIPTION
**Pull request by dpsoft**
__Improvements__
* Remove Wrappers in favor to Middlewares.

__TODO__
* Client side metrics

### CODE REVIEW

**Comment by damdev** [comment]
The code looks great! A bit tricky on the client but seems necesary

File: src/main/scala/kamon/http4s/middleware/client/ClientMiddleware.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        val clientSpan = currentContext.get(Span.ContextKey)
+
+        if (clientSpan.isEmpty()) httpService(request)
+        else {
+          val clientSpanBuilder = Kamon.buildSpan(Http4s.generateHttpClientOperationName(request))
+            .asChildOf(clientSpan)
+            .withMetricTag("span.kind", "client")
+            .withTag("http.method", request.method.name)
+            .withTag("http.url", request.uri.renderString)
+            .withTag("component", "http4s.client")
```
Comment (ivantopo) [inline src/main/scala/kamon/http4s/middleware/client/ClientMiddleware.scala]: The `component` tag should be a metric tag.

File: src/main/scala/kamon/http4s/middleware/client/ClientMiddleware.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+          val clientRequestSpan: Span = currentContext.get(SpanCustomizer.ContextKey)
+            .customize(clientSpanBuilder)
+            .start()
+
+          val newContext = currentContext.withKey(Span.ContextKey, clientRequestSpan)
+          val encodedRequest = encodeContext(newContext, request)
+
+          httpService(encodedRequest).map { response =>
+            clientRequestSpan.tag("http.status_code", response.response.status.code)
```
Comment (ivantopo) [inline src/main/scala/kamon/http4s/middleware/client/ClientMiddleware.scala]: In the akka-http module we have now (not released yet) a setting to decide whether the response status code will also be used as a metric tag, I think we should have that here as well.

**Comment by cspinetta** [comment]
@dpsoft It looks really good!! Thanks for the effort!

**Comment by dpsoft** [comment]
@saraiva132 any thoughts?

**Comment by saraiva132** [comment]
@dpsoft Looks really good. Eager to try it out.

**Comment by dpsoft** [comment]
@saraiva132  I just published a snapshot version of `kamon-http4s` https://bintray.com/kamon-io/snapshots/kamon-http4s/1.0.1-1642154649fbf48489ede49dcf411da01b8578fa in our bintray repository the middleware are `kamon.http4s.middleware.server.ServerMiddleware` and `kamon.http4s.middleware.client.ClientMiddleware`

**Comment by aeons** [comment]
I only had some issues with some of the examples in the readme iirc.

@SystemFW I think you had some more comments in the gitter channel?

**Comment by SystemFw** [comment]
Well, my comments were in the context of http4s, in which case the main issue with this code is that it's very, very side effectful. I don't know if this is a concern for this particular codebase though.

The other thing I see it's usage of things like fs2.Task, which we have removed in the newest fs2 in favour of cats effect IO, but again, this might be because this particular codebase needs to support an older version?

**Comment by cspinetta** [comment]
Hi @SystemFw! maybe you didn't see the changes introduced by this PR? @dpsoft introduced several changes related to your comments in gitter. Could you review it? Thanks!

**Comment by dpsoft** [comment]
@SystemFw thanks you for your comments!. In this PR I made several changes based on your advices in gitter, if you can please take a look to this PR.

**Comment by SystemFw** [comment]
@cspinetta @dpsoft Yeah, sorry, I just replied to @aeons message on my phone without looking at the PR, I can definitely review it if you want.
I just want to clarity that I'm not here to criticise your project :)

**Comment by SystemFw** [comment]
So, I've taken a quick look around and it looks *a lot* better, good job guys :)

There's only two bits of feedback:

- I _suspect_ some of the code in client and server can be simplified. I'm not talking about conceptual changes, just low level refactorings. This is really minor.
- There's a design choice to be made wrt to `NameGenerator`. At the moment you have global mutable shared state (`var` in an object). I can see why this might make sense at first, but in my experience using this pattern (even in Java) ultimately leads to pain and coupling. Changing this to be pure (and non-static) would however require some restructuring, so I can see why you'd want to avoid it.

Generally speaking though, I don't think any of these should stop a first release, things can always be changed/improved later (or not).

Good job!

**Comment by dpsoft** [comment]
@SystemFw Thanks for the review. We really appreciate the advices and please any comment about how simplify or change the NameGenerator are allways welcome. We will make the first release and then we try to improve the `NameGenerator` thing.

**Comment by dpsoft** [comment]
@aeons I've updated the example for one more elaborated: https://github.com/kamon-io/kamon-http4s/blob/improvements/README.md

**Comment by aeons** [comment]
Looks great 👍

**Pull request merged**